ctxhandler.go

```diff
@@ -76,52 +81,152 @@ import (
 // It uses handler returned by FromContext or fallback handler.
 // Optionally add !BADCTX attr if FromContext returns nil.
-func (*CtxHandler) Handle(context.Context, slog.Record) error {
-	panic("TODO")
+func (h *CtxHandler) Handle(ctx context.Context, r slog.Record) error {
+	handler := FromContext(ctx)
+	if handler == nil {
+		handler = h.handler
+		if !h.ignoreBADCTX {
+			handler = handler.WithAttrs([]slog.Attr{slog.Any(KeyBadCtx, ctx)})
+		}
+	}
+	for _, op := range h.opList {
+		if len(op.group) > 0 {
+			handler = handler.WithGroup(op.group)
+		} else {
+			handler = handler.WithAttrs(op.attrs)
+		}
+	}
+	return handler.Handle(ctx, r)
 }
 
 // WithAttrs works as (slog.Handler).WithAttrs.
-func (*CtxHandler) WithAttrs([]slog.Attr) slog.Handler {
-	panic("TODO")
+func (h *CtxHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
+	if len(attrs) == 0 {
+		return h
+	}
+	ctxHandler := h.clone()
+	ctxHandler.opList = append(ctxHandler.opList, data{
+		attrs: attrs,
+	})
+	return ctxHandler
 }
 
 // WithGroup works as (slog.Handler).WithGroup.
-func (*CtxHandler) WithGroup(string) slog.Handler {
-	panic("TODO")
+func (h *CtxHandler) WithGroup(name string) slog.Handler {
+	if name == "" {
+		return h
+	}
+	ctxHandler := h.clone()
+	ctxHandler.opList = append(ctxHandler.opList, data{
+		group: name,
+	})
+	return ctxHandler
 }
 
 // SetDefaultCtxHandler sets a CtxHandler as a default logger
 // and returns context with set handler inside.
 // It applies given options. If opts is nil, the default options are used.
-func SetDefaultCtxHandler(slog.Handler, ...ctxHandlerOption) context.Context {
-	panic("TODO")
+func SetDefaultCtxHandler(fallback slog.Handler, opts ...ctxHandlerOption) context.Context {
+	const size = 64 << 10
+	ctxHandler := &CtxHandler{
+		handler: fallback,
+		opList:  make([]data, size),
+	}
+	for _, opt := range opts {
+		opt(ctxHandler)
+	}
+	slog.SetDefault(slog.New(ctxHandler))
+
+	return NewContext(context.Background(), fallback)
 }
 
 // ContextWithAttrs applies attrs to a handler stored in ctx.
-func ContextWithAttrs(context.Context, ...any) context.Context {
-	panic("TODO")
+func ContextWithAttrs(ctx context.Context, attrs ...any) context.Context {
+	handler := FromContext(ctx)
+	return NewContext(ctx, handler.WithAttrs(argsToAttrSlice(attrs)))
 }
 
 // ContextWithGroup applies group to a handler stored in ctx.
-func ContextWithGroup(context.Context, string) context.Context {
-	panic("TODO")
+func ContextWithGroup(ctx context.Context, group string) context.Context {
+	handler := FromContext(ctx)
+	return NewContext(ctx, handler.WithGroup(group))
 }
 
 // LaxCtxHandler is an option for disable adding !BADCTX attr.
 func LaxCtxHandler() ctxHandlerOption { //nolint:revive // By design.
-	panic("TODO")
+	return func(ctxHandler *CtxHandler) {
+		ctxHandler.ignoreBADCTX = true
+	}
+}
+
+func (h *CtxHandler) clone() *CtxHandler {
+	return &CtxHandler{
+		handler:      h.handler,
+		opList:       h.opList,
```

Comment:
Bug: array under h.opList must be a copy (with extra capability for 1 more element).